File: src/commonTest/kotlin/com/scoreo/application/AddGameTypeUseCaseTest.kt
package com.scoreo.application

import com.scoreo.domain.DomainError
import com.scoreo.infrastructure.InMemoryGameTypeRepository
import com.scoreo.domain.model.TieBreakRule
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class AddGameTypeUseCaseTest {

    @Test
    fun `adds game type to repository`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("Belote", WinCondition.HIGHEST_SCORE)
        assertEquals(1, repo.getAll().size)
        assertEquals("Belote", repo.getAll().first().name)
    }

    @Test
    fun `trims whitespace from name`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("  Golf  ", WinCondition.LOWEST_SCORE)
        assertEquals("Golf", repo.getAll().first().name)
    }

    @Test
    fun `each game type gets a unique id`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("Belote", WinCondition.HIGHEST_SCORE)
        useCase("Golf", WinCondition.LOWEST_SCORE)
        val ids = repo.getAll().map { it.id }
        assertNotEquals(ids[0], ids[1])
    }

    @Test
    fun `returns the created game type`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        val result = useCase("Belote", WinCondition.HIGHEST_SCORE)
        assertNotNull(result.id)
        assertEquals("Belote", result.name)
        assertEquals(WinCondition.HIGHEST_SCORE, result.winCondition)
    }

    @Test
    fun `stores winCondition correctly`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("Golf", WinCondition.LOWEST_SCORE)
        assertEquals(WinCondition.LOWEST_SCORE, repo.getAll().first().winCondition)
    }

    @Test
    fun `multiple game types are all saved`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("A", WinCondition.HIGHEST_SCORE)
        useCase("B", WinCondition.LOWEST_SCORE)
        useCase("C", WinCondition.MANUAL)
        assertEquals(3, repo.getAll().size)
    }

    @Test
    fun `rejects name exceeding 50 characters`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        val longName = "A".repeat(51)
        val exception = assertFailsWith<DomainError.Validation> {
            useCase(longName, WinCondition.HIGHEST_SCORE)
        }
        assertTrue(exception.message!!.contains("50"))
    }

    @Test
    fun `accepts name of exactly 50 characters`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        val name = "A".repeat(50)
        val gameType = useCase(name, WinCondition.HIGHEST_SCORE)
        assertEquals(50, gameType.name.length)
    }

    // P0-04: Tie Break Configuration Tests
    @Test
    fun `stores tie break rule correctly`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase(
            "Golf",
            WinCondition.HIGHEST_SCORE,
            tieBreakRule = TieBreakRule.SECONDARY_SCORE,
        )
        val gameType = repo.getAll().first()
        assertEquals(TieBreakRule.SECONDARY_SCORE, gameType.tieBreakRule)
    }

    @Test
    fun `stores tie break condition correctly`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase(
            "Golf",
            WinCondition.HIGHEST_SCORE,
            tieBreakRule = TieBreakRule.SECONDARY_SCORE,
            tieBreakCondition = WinCondition.LOWEST_SCORE,
        )
        val gameType = repo.getAll().first()
        assertEquals(WinCondition.LOWEST_SCORE, gameType.tieBreakCondition)
    }

    @Test
    fun `stores tie break label correctly`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase(
            "Golf",
            WinCondition.HIGHEST_SCORE,
            tieBreakRule = TieBreakRule.SECONDARY_SCORE,
            tieBreakCondition = WinCondition.LOWEST_SCORE,
            tieBreakLabel = "Handicap",
        )
        val gameType = repo.getAll().first()
        assertEquals("Handicap", gameType.tieBreakLabel)
    }

    @Test
    fun `defaults tie break rule to NONE`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("Belote", WinCondition.HIGHEST_SCORE)
        assertEquals(TieBreakRule.NONE, repo.getAll().first().tieBreakRule)
    }

    @Test
    fun `defaults tie break label to null`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = AddGameTypeUseCase(repo)
        useCase("Belote", WinCondition.HIGHEST_SCORE)
        assertNull(repo.getAll().first().tieBreakLabel)
    }
}

File: apps/scoreo/src/application/addGameTypeUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { ValidationError } from '../domain/model/errors'
import { InMemoryGameTypeRepository } from '../infrastructure/testing/inMemoryGameTypeRepository'
import { AddGameTypeUseCase } from './addGameTypeUseCase'

describe('AddGameTypeUseCase', () => {
  it('adds game type to repository', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Belote', 'HIGHEST_SCORE')
    expect(repo.getAll()).toHaveLength(1)
    expect(repo.getAll()[0].name).toBe('Belote')
  })

  it('trims whitespace from name', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('  Golf  ', 'LOWEST_SCORE')
    expect(repo.getAll()[0].name).toBe('Golf')
  })

  it('each game type gets a unique id', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Belote', 'HIGHEST_SCORE')
    useCase.invoke('Golf', 'LOWEST_SCORE')
    const ids = repo.getAll().map((g) => g.id)
    expect(ids[0]).not.toBe(ids[1])
  })

  it('returns the created game type', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    const result = useCase.invoke('Belote', 'HIGHEST_SCORE')
    expect(result.id).toBeTruthy()
    expect(result.name).toBe('Belote')
    expect(result.winCondition).toBe('HIGHEST_SCORE')
  })

  it('stores winCondition correctly', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Golf', 'LOWEST_SCORE')
    expect(repo.getAll()[0].winCondition).toBe('LOWEST_SCORE')
  })

  it('multiple game types are all saved', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('A', 'HIGHEST_SCORE')
    useCase.invoke('B', 'LOWEST_SCORE')
    useCase.invoke('C', 'MANUAL')
    expect(repo.getAll()).toHaveLength(3)
  })

  it('rejects name exceeding 50 characters', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    const longName = 'A'.repeat(51)
    let error: unknown
    try {
      useCase.invoke(longName, 'HIGHEST_SCORE')
    } catch (e) {
      error = e
    }
    expect(error).toBeInstanceOf(ValidationError)
    expect((error as ValidationError).message).toContain('50')
  })

  it('accepts name of exactly 50 characters', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    const name = 'A'.repeat(50)
    const gameType = useCase.invoke(name, 'HIGHEST_SCORE')
    expect(gameType.name).toHaveLength(50)
  })

  it('stores tie break rule correctly', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Golf', 'HIGHEST_SCORE', { tieBreakRule: 'SECONDARY_SCORE' })
    expect(repo.getAll()[0].tieBreakRule).toBe('SECONDARY_SCORE')
  })

  it('stores tie break condition correctly', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Golf', 'HIGHEST_SCORE', {
      tieBreakRule: 'SECONDARY_SCORE',
      tieBreakCondition: 'LOWEST_SCORE',
    })
    expect(repo.getAll()[0].tieBreakCondition).toBe('LOWEST_SCORE')
  })

  it('stores tie break label correctly', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Golf', 'HIGHEST_SCORE', {
      tieBreakRule: 'SECONDARY_SCORE',
      tieBreakCondition: 'LOWEST_SCORE',
      tieBreakLabel: 'Handicap',
    })
    expect(repo.getAll()[0].tieBreakLabel).toBe('Handicap')
  })

  it('defaults tie break rule to NONE', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Belote', 'HIGHEST_SCORE')
    expect(repo.getAll()[0].tieBreakRule).toBe('NONE')
  })

  it('defaults tie break label to null', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new AddGameTypeUseCase(repo)
    useCase.invoke('Belote', 'HIGHEST_SCORE')
    expect(repo.getAll()[0].tieBreakLabel).toBeNull()
  })
})
